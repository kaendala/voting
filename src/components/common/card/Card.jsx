import "./Card.scss";
import React, { useEffect, useRef, useState } from "react";

const Card = ({ person, sort, changeVotes }) => {
  const [currentVote, setCurrentVote] = useState(null);
  const [picture, setPicture] = useState(null);
  const [positivePorcent, setPositivePorcent] = useState(null);
  const [negativePorcent, setNegativePorcent] = useState(null);
  const [optionSelect, setOptionSelect] = useState(null);
  const positive = useRef(null);
  const negative = useRef(null);

  useEffect(() => {
    const positiveVotes = person.votes.positive;
    const negativeVotes = person.votes.negative;
    const total = positiveVotes + negativeVotes;
    const posPorcent = ((positiveVotes * 100) / total).toFixed(1) + "%";
    const negPorcent = ((negativeVotes * 100) / total).toFixed(1) + "%";
    positive.current.style.width = posPorcent;
    negative.current.style.width = negPorcent;
    setNegativePorcent(negPorcent);
    setPositivePorcent(posPorcent);
  }, [person]);
  useEffect(() => {
    let picture = person.picture;
    if (sort === "grid") {
      picture = picture.split(".");
      picture = picture[0] + "G." + picture[1];
    }
    setPicture(picture);
  }, [sort, person.picture]);
  const difDate = (last) => {
    const difdt = new Date(new Date() - new Date(last));
    let date = "";
    date =
      difdt.toISOString().slice(0, 4) - 1970 === 0
        ? ""
        : difdt.toISOString().slice(0, 4) - 1970 + " year ago";
    if (!date) {
      date = date === "" ? difdt.getMonth() + 1 + " months ago" : "";
    }
    if (!date) {
      date = date === "" ? difdt.getDate() + " days ago" : "";
    }
    return date;
  };
  const hand = (background, position) => {
    return (
      <div
        className={
          "hand" +
          (background ? " background" : "") +
          (position === "up" ? " up" : " down")
        }
      >
        <img
          src={"http://localhost:3000/assets/img/thumbs-up.svg"}
          alt="hand"
        ></img>
      </div>
    );
  };
  return (
    <div className={"card " + sort}>
      {hand(
        true,
        person.votes.positive > person.votes.negative ? "up" : "down"
      )}
      <div className="picture">
        <img alt="picturePerson" src={`/assets/img/people/${picture}`}></img>
      </div>
      <div className="description">
        <div className="contentName">
          <h2 className="name">{person.name}</h2>
          <h4 className="info">{person.description}</h4>
        </div>
        <div className={"votes " + (currentVote ? "vote" : "noVote")}>
          <div className="lastUpdated">
            <h5>
              {currentVote
                ? "Thank you for your vote!"
                : difDate(person.lastUpdated) + " in " + person.category}
            </h5>
          </div>
          <div className="buttons">
            <div
              className={optionSelect === "positive" ? " selected" : ""}
              onClick={() => {
                setOptionSelect("positive");
              }}
            >
              {hand(true, "up")}
            </div>
            <div
              className={optionSelect === "negative" ? " selected" : ""}
              onClick={() => {
                setOptionSelect("negative");
              }}
            >
              {hand(true, "down")}
            </div>
            <button
              disabled={!optionSelect && !currentVote}
              onClick={() => {
                if (currentVote) {
                  setCurrentVote(false);
                } else {
                  changeVotes(optionSelect, person);
                  setCurrentVote(true);
                  setOptionSelect(null);
                }
              }}
            >
              {currentVote ? "Vote Again" : "Vote Now"}
            </button>
          </div>
        </div>
      </div>
      <div className="statistics">
        <div ref={positive} className="statistic positive">
          {hand(false, "up")}
          <h3>{positivePorcent}</h3>
        </div>
        <div ref={negative} className="statistic negative">
          <h3>{negativePorcent}</h3>
          {hand(false, "down")}
        </div>
      </div>
    </div>
  );
};

export default Card;
