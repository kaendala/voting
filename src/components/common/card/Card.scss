.card {
  position: relative;
  .hand {
    position: absolute;
    top: 0;
    z-index: 1;
    display: flex;
    align-items: center;
    justify-content: center;
    &.background.down {
      background-color: rgba(var(--color-yellow-negative), 0.8);
    }
    &.background.up {
      background-color: rgba(var(--color-green-positive), 0.8);
    }
    &.down {
      transform: rotate(180deg);
    }
    img {
      width: 53.4%;
    }
  }
  .picture {
    position: absolute;
    height: 100%;
    img {
      width: 100%;
      height: 100%;
      object-fit: cover;
    }
  }
  .description {
    color: var(--color-white);
    position: absolute;
    display: flex;
    background: linear-gradient(
      90deg,
      rgba(0, 0, 0, 0.0001) 0%,
      #888888 19.79%,
      #666666 50%,
      rgba(51, 51, 51, 0.6) 71.88%
    );
    .name,
    .info {
      overflow: hidden;
      text-overflow: ellipsis;
      display: -webkit-box;
      -webkit-line-clamp: 2;
      -webkit-box-orient: vertical;
    }
    .votes {
      .lastUpdated {
        text-align: right;
      }
      .buttons {
        display: flex;
        justify-content: space-between;
        align-items: center;
        .hand {
          top: auto;
          position: relative;
        }
        button {
          background-color: rgba(48, 48, 48, 0.6);
          color: var(--color-white);
          border: 1px solid var(--color-white);
          cursor: pointer;
          &:disabled {
            background-color: var(--color-darker-background);
            cursor: not-allowed;
          }
        }
        .selected {
          border: 1px solid var(--color-white);
        }
      }
      &.vote {
        .hand {
          visibility: hidden;
        }
      }
    }
  }
  &.list {
    height: 170px;
    width: 100%;
    margin-bottom: 18px;
    .hand {
      width: 45px;
      height: 45px;
    }
    .picture {
      width: 24.28%;
    }
    .description {
      height: 100%;
      left: 4.15%;
      padding-left: 19.37%;
      width: 76.5%;
      .contentName {
        width: 69.3%;
        .info {
          width: 92.7%;
        }
      }
      .votes {
        width: 28.9%;
        padding-right: 14px;
        .lastUpdated {
          margin-top: 8px;
        }
        .buttons {
          margin-top: 17px;

          button {
            width: 135px;
            height: 45px;
          }
        }
      }
    }
    .statistics {
      height: 54px;
    }
  }
  &.grid {
    height: 348px;
    width: 31.64%;
    margin-bottom: 27px;
    .hand {
      width: 30px;
      height: 30px;
      top: 41.1%;
    }
    .description {
      display: block;
      height: 71.4%;
      background: linear-gradient(
        180deg,
        rgba(0, 0, 0, 0.0001) 0%,
        rgba(0, 0, 0, 0.6) 100%
      );
      bottom: 0;
      .contentName {
        padding-left: 36px;
        margin-top: 24%;
        width: 85%;
        position: relative;
        .name {
          line-height: 43px;
          position: absolute;
          bottom: 0;
          margin-bottom: 11%;
        }
      }
      .votes {
        padding-right: 9.8%;
        width: 44%;
        position: absolute;
        right: 0;
        .lastUpdated {
          margin-bottom: 12px;
          h5 {
            font-size: 12px;
            line-height: 14.4px;
            margin-top: 12px;
          }
        }
        .buttons {
          align-items: center;
          button {
            height: 36px;
          }
        }
      }
    }
    h4 {
      font-size: 15px;
      line-height: 18px;
    }
    .statistics {
      height: 36px;
      h3 {
        font-size: 18px;
        line-height: 21.6px;
      }
    }
  }
  .statistics {
    position: absolute;
    display: flex;
    bottom: 0;
    width: 100%;
    .statistic {
      display: flex;
      align-items: center;
      color: var(--color-white);
      &.positive {
        background-color: rgba(var(--color-green-positive), 0.6);
        height: 100%;
      }
      &.negative {
        background-color: rgba(var(--color-yellow-negative), 0.6);
        height: 100%;
        justify-content: flex-end;
      }
      .hand {
        position: initial;
        img {
          width: 50%;
        }
      }
    }
  }
}
@media (max-width: 1024px) {
  .card {
    &.list {
      height: 142px;
      .hand {
        width: 30px;
        height: 30px;
      }
      .description {
        .contentName {
          .name {
            margin-bottom: 18px;
          }
        }
        .name {
          -webkit-line-clamp: 1;
        }
        .votes {
          width: 40%;
          max-width: 202px;
          .lastUpdated {
            h5 {
              font-weight: 700;
            }
          }
          .buttons {
            button {
              height: 38px;
              width: 107px;
            }
          }
        }
      }
      .statistics {
        height: 36px;
      }
    }
    &.grid {
      width: 48.9%;
      .picture {
        width: 100%;
      }
      .description {
        .votes {
          width: 54%;
          max-width: 200px;
        }
        .contentName {
          margin-top: 85px;
        }
      }
    }
  }
}
@media (max-width: 767px) {
  .card {
    &.grid {
      width: 100%;
      height: 300px;
      margin-bottom: 0;
      .description {
        height: 85%;
        .contentName {
          .name {
            margin-bottom: 36px;
            line-height: 36px;
          }
        }
      }
      .hand {
        top: 31%;
      }
    }
  }
}
