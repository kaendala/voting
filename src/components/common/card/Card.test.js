import React from 'react';
import Card from './Card';
import { data } from '../../../../public/assets/data.json'
import { render } from '@testing-library/react';

describe('Test for Card component', () => {
    let component;
    let vote=false
    beforeEach(() => {
         vote=false
    })
    test('should load component Card sort list ', () => {
        component = render(<Card person={data[0]} sort="list" changeVotes={changeVotes}></Card>);
        expect(component).toMatchSnapshot();
    });
    test('should load component Card  sort grid', () => {
        component = render(<Card person={data[3]} sort="grid" changeVotes={changeVotes}></Card>);
        expect(component).toMatchSnapshot();
    });
    test('should load call change votes positive', () => {
        component = render(<Card person={data[2]} sort="grid" changeVotes={changeVotes}></Card>);
        component.container.querySelector(".buttons div").click()
        component.container.querySelector(".buttons button").click()
        component.container.querySelector(".buttons button").click()
        expect(vote).toBe(true);
    });
    test('should load call change votes negative', () => {
        component = render(<Card person={data[5]} sort="list" changeVotes={changeVotes}></Card>);
        component.container.querySelectorAll(".buttons div")[2].click()
        component.container.querySelector(".buttons button").click()
        expect(vote).toBe(true);
    });
    const changeVotes=() => {
        vote=true
    }

})