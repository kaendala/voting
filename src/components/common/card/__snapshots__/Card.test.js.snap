// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Card component should load component Card  sort grid 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card grid"
      >
        <div
          class="hand background up"
        >
          <img
            alt="hand"
            src="http://localhost:3000/assets/img/thumbs-up.svg"
          />
        </div>
        <div
          class="picture"
        >
          <img
            alt="picturePerson"
            src="/assets/img/people/malalaG.png"
          />
        </div>
        <div
          class="description"
        >
          <div
            class="contentName"
          >
            <h2
              class="name"
            >
              Malala Yousafzai
            </h2>
            <h4
              class="info"
            >
              The daughter of educational activist Ziauddin, Yousafzai was born to a Pashtun family in Mingora, Khyber Pakhtunkhwa, Pakistan. Her family came to run a chain of schools in the region.
            </h4>
          </div>
          <div
            class="votes noVote"
          >
            <div
              class="lastUpdated"
            >
              <h5>
                8 months ago in politics
              </h5>
            </div>
            <div
              class="buttons"
            >
              <div
                class=""
              >
                <div
                  class="hand background up"
                >
                  <img
                    alt="hand"
                    src="http://localhost:3000/assets/img/thumbs-up.svg"
                  />
                </div>
              </div>
              <div
                class=""
              >
                <div
                  class="hand background down"
                >
                  <img
                    alt="hand"
                    src="http://localhost:3000/assets/img/thumbs-up.svg"
                  />
                </div>
              </div>
              <button
                disabled=""
              >
                Vote Now
              </button>
            </div>
          </div>
        </div>
        <div
          class="statistics"
        >
          <div
            class="statistic positive"
            style="width: 85.7%;"
          >
            <div
              class="hand up"
            >
              <img
                alt="hand"
                src="http://localhost:3000/assets/img/thumbs-up.svg"
              />
            </div>
            <h3>
              85.7%
            </h3>
          </div>
          <div
            class="statistic negative"
            style="width: 14.3%;"
          >
            <h3>
              14.3%
            </h3>
            <div
              class="hand down"
            >
              <img
                alt="hand"
                src="http://localhost:3000/assets/img/thumbs-up.svg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card grid"
    >
      <div
        class="hand background up"
      >
        <img
          alt="hand"
          src="http://localhost:3000/assets/img/thumbs-up.svg"
        />
      </div>
      <div
        class="picture"
      >
        <img
          alt="picturePerson"
          src="/assets/img/people/malalaG.png"
        />
      </div>
      <div
        class="description"
      >
        <div
          class="contentName"
        >
          <h2
            class="name"
          >
            Malala Yousafzai
          </h2>
          <h4
            class="info"
          >
            The daughter of educational activist Ziauddin, Yousafzai was born to a Pashtun family in Mingora, Khyber Pakhtunkhwa, Pakistan. Her family came to run a chain of schools in the region.
          </h4>
        </div>
        <div
          class="votes noVote"
        >
          <div
            class="lastUpdated"
          >
            <h5>
              8 months ago in politics
            </h5>
          </div>
          <div
            class="buttons"
          >
            <div
              class=""
            >
              <div
                class="hand background up"
              >
                <img
                  alt="hand"
                  src="http://localhost:3000/assets/img/thumbs-up.svg"
                />
              </div>
            </div>
            <div
              class=""
            >
              <div
                class="hand background down"
              >
                <img
                  alt="hand"
                  src="http://localhost:3000/assets/img/thumbs-up.svg"
                />
              </div>
            </div>
            <button
              disabled=""
            >
              Vote Now
            </button>
          </div>
        </div>
      </div>
      <div
        class="statistics"
      >
        <div
          class="statistic positive"
          style="width: 85.7%;"
        >
          <div
            class="hand up"
          >
            <img
              alt="hand"
              src="http://localhost:3000/assets/img/thumbs-up.svg"
            />
          </div>
          <h3>
            85.7%
          </h3>
        </div>
        <div
          class="statistic negative"
          style="width: 14.3%;"
        >
          <h3>
            14.3%
          </h3>
          <div
            class="hand down"
          >
            <img
              alt="hand"
              src="http://localhost:3000/assets/img/thumbs-up.svg"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test for Card component should load component Card sort list  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card list"
      >
        <div
          class="hand background down"
        >
          <img
            alt="hand"
            src="http://localhost:3000/assets/img/thumbs-up.svg"
          />
        </div>
        <div
          class="picture"
        >
          <img
            alt="picturePerson"
            src="/assets/img/people/kanye.png"
          />
        </div>
        <div
          class="description"
        >
          <div
            class="contentName"
          >
            <h2
              class="name"
            >
              Kanye West
            </h2>
            <h4
              class="info"
            >
              Born in Atlanta and raised in Chicago, West was first known as a producer for Roc-A-Fella Records in the early 2000s, producing singles for several mainstream artists.
            </h4>
          </div>
          <div
            class="votes noVote"
          >
            <div
              class="lastUpdated"
            >
              <h5>
                1 year ago in entertainment
              </h5>
            </div>
            <div
              class="buttons"
            >
              <div
                class=""
              >
                <div
                  class="hand background up"
                >
                  <img
                    alt="hand"
                    src="http://localhost:3000/assets/img/thumbs-up.svg"
                  />
                </div>
              </div>
              <div
                class=""
              >
                <div
                  class="hand background down"
                >
                  <img
                    alt="hand"
                    src="http://localhost:3000/assets/img/thumbs-up.svg"
                  />
                </div>
              </div>
              <button
                disabled=""
              >
                Vote Now
              </button>
            </div>
          </div>
        </div>
        <div
          class="statistics"
        >
          <div
            class="statistic positive"
            style="width: 39.0%;"
          >
            <div
              class="hand up"
            >
              <img
                alt="hand"
                src="http://localhost:3000/assets/img/thumbs-up.svg"
              />
            </div>
            <h3>
              39.0%
            </h3>
          </div>
          <div
            class="statistic negative"
            style="width: 61.0%;"
          >
            <h3>
              61.0%
            </h3>
            <div
              class="hand down"
            >
              <img
                alt="hand"
                src="http://localhost:3000/assets/img/thumbs-up.svg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card list"
    >
      <div
        class="hand background down"
      >
        <img
          alt="hand"
          src="http://localhost:3000/assets/img/thumbs-up.svg"
        />
      </div>
      <div
        class="picture"
      >
        <img
          alt="picturePerson"
          src="/assets/img/people/kanye.png"
        />
      </div>
      <div
        class="description"
      >
        <div
          class="contentName"
        >
          <h2
            class="name"
          >
            Kanye West
          </h2>
          <h4
            class="info"
          >
            Born in Atlanta and raised in Chicago, West was first known as a producer for Roc-A-Fella Records in the early 2000s, producing singles for several mainstream artists.
          </h4>
        </div>
        <div
          class="votes noVote"
        >
          <div
            class="lastUpdated"
          >
            <h5>
              1 year ago in entertainment
            </h5>
          </div>
          <div
            class="buttons"
          >
            <div
              class=""
            >
              <div
                class="hand background up"
              >
                <img
                  alt="hand"
                  src="http://localhost:3000/assets/img/thumbs-up.svg"
                />
              </div>
            </div>
            <div
              class=""
            >
              <div
                class="hand background down"
              >
                <img
                  alt="hand"
                  src="http://localhost:3000/assets/img/thumbs-up.svg"
                />
              </div>
            </div>
            <button
              disabled=""
            >
              Vote Now
            </button>
          </div>
        </div>
      </div>
      <div
        class="statistics"
      >
        <div
          class="statistic positive"
          style="width: 39.0%;"
        >
          <div
            class="hand up"
          >
            <img
              alt="hand"
              src="http://localhost:3000/assets/img/thumbs-up.svg"
            />
          </div>
          <h3>
            39.0%
          </h3>
        </div>
        <div
          class="statistic negative"
          style="width: 61.0%;"
        >
          <h3>
            61.0%
          </h3>
          <div
            class="hand down"
          >
            <img
              alt="hand"
              src="http://localhost:3000/assets/img/thumbs-up.svg"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
